### PR TITLE
Kentik custom dimensions support #46

### PR DESCRIPTION
**Pull request by rozetko**
Closes <placeholder>

- Add custom dimensions to "Group by"
- Add custom dimensions to "Filters"

image
image
image

### CODE REVIEW

File: src/datasource/datasource.ts [hunk 1]
```diff
@@ -74,7 +75,11 @@ class KentikDatasource {
       return [];
     }
 
-    const metricDef = _.find(metricList, { value: query.dimension[0] });
+    const metricDef = _.find(
+      await this._getExtendedDimensionsList(metricList),
```
Comment (jonyrock) [inline src/datasource/datasource.ts]: Please assign `await this._getExtendedDimensionsList(metricList)` and use in function

File: src/datasource/datasource.ts [hunk 2]
```diff
@@ -143,7 +148,7 @@ class KentikDatasource {
 
   metricFindQuery(query) {
     if (query === 'metrics()') {
-      return Promise.resolve(metricList);
+      return Promise.resolve(this._getExtendedDimensionsList(metricList));
```
Comment (jonyrock) [inline src/datasource/datasource.ts]: just `return this._getExtendedDimensionsList(metricList)` and make `async metricFindQuery`

File: src/datasource/datasource.ts [hunk 3]
```diff
@@ -157,21 +162,34 @@ class KentikDatasource {
   }
 
   getTagKeys() {
-    return Promise.resolve(filterFieldList);
+    return Promise.resolve(this._getExtendedDimensionsList(filterFieldList));
```
Comment (jonyrock) [inline src/datasource/datasource.ts]: same shit as with `_getExtendedDimensionsList`

File: src/datasource/datasource.ts [hunk 4]
```diff
@@ -157,21 +162,34 @@ class KentikDatasource {
...
+        const field = filter.field;
+        return this.kentik.getFieldValues(field).then(result => {
+          return result.rows.map(row => {
+            return { text: row[field].toString() };
+          });
         });
-      });
+      }
     } else {
       return Promise.resolve([]);
```
Comment (jonyrock) [inline src/datasource/datasource.ts]: if it is `async`, you can just return `[]`

File: src/datasource/kentikAPI.ts [hunk 5]
```diff
@@ -23,6 +23,11 @@ class KentikAPI {
     return this.invokeSQLQuery(query);
   }
 
+  getCustomDimensions() {
+    return this._get('/api/v5/customdimensions')
```
Comment (jonyrock) [inline src/datasource/kentikAPI.ts]:
```
async getCustomDimensions() {
  const data = await this._get('/api/v5/customdimensions');
  return data.data.customDimensions;
}
```

Comment (jonyrock) [reply src/datasource/kentikAPI.ts]: It's because we use `async/await` eveywhere

File: src/datasource/kentikProxy.ts [hunk 6]
```diff
@@ -113,6 +113,26 @@ class KentikProxy {
       });
     }
   }
+
+  async getCustomDimensions() {
+    if (this.cache.customDimensions === undefined) {
+      const customDimensions = await this.kentikAPI.getCustomDimensions();
+      this.cache.customDimensions = customDimensions.map(dimension => {
```
Comment (jonyrock) [inline src/datasource/kentikProxy.ts]: In `map(dimension => <code>)`  
lets move `<code>` to a function. Hard to read this.

File: src/datasource/query_builder.ts [hunk 7]
```diff
@@ -123,15 +123,16 @@ function buildTopXdataQuery(options) {
   return query;
 }
 
-function convertToKentikFilter(filterObj) {
+function convertToKentikFilter(filterObj, customDimensions: Array<any>) {
```
Comment (jonyrock) [inline src/datasource/query_builder.ts]: we use `Array<any>`, not `any[]`, right ?

File: src/datasource/query_builder.ts [hunk 8]
```diff
@@ -145,9 +146,9 @@ function convertToKentikFilter(filterObj) {
   };
 }
 
-function convertToKentikFilterGroup(filters) {
+function convertToKentikFilterGroup(filters, customDimensions: Array<any>) {
```
Comment (jonyrock) [inline src/datasource/query_builder.ts]: Same qestoin: we use `Array<any>`, not `any[]`, right ?

**Comment by rozetko** [comment]
Created PR to `grafana:master`: https://github.com/grafana/kentik-app/pull/52

**Pull request closed without merging**